src/core/seat/input-method-relay.cpp

```diff
@@ -100,6 +100,7 @@ wf::input_method_relay::input_method_relay()
         on_new_popup_surface.disconnect();
         input_method  = nullptr;
         keyboard_grab = nullptr;
+        last_keyboard_resource = nullptr;
```

Comment:
`keyboard_grab` will be null on unfocus, so I keep `keyboard_grab->resource` for comparison.

When the input method also provide an OSK, we can't distinguish if the input method is passing a user-pressed key back, or the user presses the OSK provided by the input method.

Reply:
* create a window and activate IM
* open another window
* close the window with keyboard (e.g. super+c from wayfire, or ctrl+d for a terminal)

Some key release events will be received by the first window twice (if we return false in `is_im_sent` in case `input_method->keyboard_grab` is null). It doesn't seem to cause problems though.

Reply:
Where do the releasing the key twice come from? Once from the IM and once from the physical release, or am I missing something?

By the way, https://github.com/WayfireWM/wayfire/blob/master/src/core/seat/keyboard.cpp#L40 - is this maybe the reason why it works in practice?

Reply:
> Once from the IM and once from the physical release

Yes.

> https://github.com/WayfireWM/wayfire/blob/master/src/core/seat/keyboard.cpp#L40 - is this maybe the reason why it works in practice?

It seems so. If the events aren't sent to client twice, it shouldn't matter. But I prefer to keep the processing in the same place and make the `is_im_sent` method accurate.

Reply:
Alright, I see your point. Now on to the technical realization: I see that you are storing keyboard_grab->resource. Is it not technically possible that the resource is destroyed but we still keep holding a reference to it? `last_keyboard_resource` seems to be reset on input-method-destroy, which to my understanding can be also happening at a much later point than grab destroy.

Another thing: when you call wl_resource_get_client() in `is_im_sent`, if the IM has not used a grab,  the chosen resource will be NULL, right? Wouldn't that lead to a crash when you do a wl_resource_get_client()?

Last but not least, in the current implementation, a grab is grabbed forever (in the sense that the keys will never reach the clients, because is_im_sent returns true) - because last_keyboard_resource is cleared only when the whole IM is destroyed, right? This is a protocol violation to my understanding.

I think in general it might make much more sense to just filter out the few release events in the end (like we already do in core, if that is not enough, feel free to do something similar in the input method relay).

Reply:
> Is it not technically possible that the resource is destroyed but we still keep holding a reference to it?

I think so, but it should be fine as long as it isn't dereferenced.

> when you call wl_resource_get_client() in is_im_sent, if the IM has not used a grab, the chosen resource will be NULL, right?

`virtual_keyboard` should also be null in this case, so we've already returned early. I can add a null check just in case.

> a grab is grabbed forever (in the sense that the keys will never reach the clients, because is_im_sent returns true) - because last_keyboard_resource is cleared only when the whole IM is destroyed, right?

`is_im_sent` only returns true when the key events are from the IM. These keys will never reach the IM or the compositor's shortcut keys logic, but they will reach the clients using text inputs.

The IM needs to bypass key events, and it does so by sending it via a virtual keyboard. We need to ensure that these events aren't sent to the IM again and get through to the text input clients. This roundabout is caused by the deficiencies of the input method v2 protocol.

Reply:
> `virtual_keyboard` should also be null in this case, so we've already returned early. I can add a null check just in case.

This is not necessary. You can have a standalone virtual keyboard which is not connected to an IM or whatever. see for example wf-osk, or wayvnc.

> I think so, but it should be fine as long as it isn't dereferenced.

Sure, but it is dereferenced when we try to get its client.

To check both cases, simply start a virtual-keyboard-only client like wf-osk and you will certainly get a crash (at least if you run with address sanitizer). I just checked :)

> is_im_sent only returns true when the key events are from the IM. These keys will never reach the IM or the compositor's shortcut keys logic, but they will reach the clients using text inputs.

But these keys will not reach applications if they are coming from a virtual keyboard, but not from the IM after the grab has ended.

Reply:
Yes, you're right. I'm switching to use pid to identify the input method.